mathphys/functions.py

```diff
@@ -197,3 +286,60 @@ def get_package_string(package):
     else:
         repo_str += f'{ver:s}'
     return repo_str
+
+
+# ------------------------- HELPER METHODS ------------------------------------
+_TYPES2SAVE = (int, float, complex, str, bytes, bool)
+_STRTYPES = {typ.__name__ for typ in _TYPES2SAVE}
+_NPTYPES = (_np.int_, _np.float_, _np.complex_, _np.bool_)
+
+
+def _save_recursive_hdf5(fil, path, obj, compress):
+    typ = type(obj)
+    if isinstance(obj, _np.ndarray):
+        fil.create_dataset(path, data=obj, compression=compress)
+    elif isinstance(obj, str):  # Handle SiriusPVName
+        fil[path] = str(obj)
+    elif isinstance(obj, _TYPES2SAVE + _NPTYPES):
+        fil[path] = obj
+    elif obj is None:
+        fil[path] = 'None'
+    elif typ in {list, tuple, set}:
+        path = path.strip('/') + '/'
+        for i, el in enumerate(obj):
+            _save_recursive_hdf5(fil, path + f'{i:d}', el, compress)
+    elif typ == dict:
+        path = path.strip('/') + '/'
+        for key, item in obj.items():
+            _save_recursive_hdf5(fil, path + key, item, compress)
+    else:
+        raise TypeError('Data type not valid: '+typ.__name__+'.')
+    fil[path].attrs['type'] = typ.__name__
+
+
+def _load_recursive_hdf5(fil):
+    """."""
+    typ = fil.attrs['type']
+    if typ == 'list':
+        return [_load_recursive_hdf5(fil[str(i)]) for i in range(len(fil))]
+    elif typ == 'tuple':
+        return tuple([
+            _load_recursive_hdf5(fil[str(i)]) for i in range(len(fil))])
+    elif typ == 'set':
+        return {_load_recursive_hdf5(fil[str(i)]) for i in range(len(fil))}
+    elif typ == 'dict':
+        return {k: _load_recursive_hdf5(obj) for k, obj in fil.items()}
+    elif typ == 'NoneType':
+        return None
+    elif typ == 'ndarray':
+        return fil[()]
+    elif typ == 'str':
+        return fil[()].decode()
+    elif _HASSIRIUSPY and typ == 'SiriusPVName':
+        return _SiriusPVName(fil[()].decode())
+    elif typ in _STRTYPES:
+        exec('type_ = '+typ)
+        return locals()['type_'](fil[()])
+    else:
+        exec('type_ = _np.'+typ)
+        return locals()['type_'](fil[()])
```

Comment:
I know this is likely to only be run for "trusted" files, but the usage of `exec` bothers me a bit from a security standpoint, though I'm not sure how to improve it...
Could you use `globals()[typ]` instead and avoid the exec?

I think I'd also tighten up the conditional (`_STRNPTYPES` would have to be defined, and we could also add an exception in case the type really is unknown):

```suggestion
    elif typ in _STRTYPES:
        exec('type_ = '+typ)
        return locals()['type_'](fil[()])
    elif typ in _STRNPTYPES:
        exec('type_ = _np.'+typ)
        return locals()['type_'](fil[()])
```

Reply:
Yeah, I also don't like this. I have a verbose way of removing the need for the exec method. I'll implement it. Thanks!

Reply:
> Yeah, I also don't like this. I have a verbose way of removing the need for the exec method. I'll implement it. Thanks!

Thinking about it, we are still executing functions from `locals()`, which isn't the safest thing, especially when it's not restricted like for `_STRTYPES`.

I was testing out what h5py does, and it seems to save the type in a `.dtype` member of the dataset, which is different from the underlying `type` of the data (`type` seems to be optimized for storage, which I think is what happened to make the data smaller: if the max of an array fits under fewer bits, it will take advantage of it). This makes it so values stored from `np.{uint,int,float,complex,bool}{8,16,32,64,128}` (btw, `uint` is missing from `_NPTYPES`, unless it's a subclass of `int`?) are restored to the original np type if you use the `[()]` syntax; so I think you can skip the function call entirely, and simply return the value like you do for `ndarrays`. For the python `int` and `float` values, it converts to `np.int64` and `np.float64`, respectively, so it seems to still make sense to call their respective functions.

> Yeah, I also don't like this. I have a verbose way of removing the need for the exec method. I'll implement it. Thanks!

Were you thinking about something like a dictionary with the functions?